In [ ]:
# Connecting to Google drive

from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


## **Data Understanding**

In [ ]:
import pandas as pd
import warnings
warnings.filterwarnings("ignore")

In [ ]:
# Loading the dataset

file_path = "/content/drive/MyDrive/CRB_Data"

application_data= pd.read_csv(file_path + "/1. ApplicationData.csv")
#/content/drive/MyDrive/CRB_Data/1. ApplicationData.csv

application_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9492 entries, 0 to 9491
Data columns (total 16 columns):
 #   Column                                         Non-Null Count  Dtype  
---  ------                                         --------------  -----  
 0   Application_ID                                 9492 non-null   object 
 1   Application_Date                               9492 non-null   object 
 2   Loan_Approval_Status                           9354 non-null   object 
 3   Age_at_Application                             9471 non-null   float64
 4   Marital_Status                                 7712 non-null   object 
 5   Gender                                         9488 non-null   object 
 6   Own_Rent_Home                                  8188 non-null   object 
 7   Occupation_Professional_Category               9361 non-null   object 
 8   Type_of_Business_Industry_of_Employment        7210 non-null   object 
 9   Employment_Status                              9357 

**Observations:**

1. There are 9492 records of the data.

2. There are 16 features: Application_ID, Application_Date, Loan_Approval_Status, Age_at_Application, Marital_Status, Gender, Own_Rent_Home,
Occupation_Professional_Category,Type_of_Business_Industry_of_Employment, Employment_Status, Time_at_Current_Employment_(Months),Time_Employed_with_Previous_Employer_(Months), Number_of_Dependents Loan_Purpose, Monthly_Income, Government_Employee

3. The data types are: float64(5), object(11)

## **ContractsData**

In [ ]:
contracts_data = pd.read_csv(file_path + "/2. ContractsData.csv")
contracts_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9374 entries, 0 to 9373
Data columns (total 9 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   Contract_ID                 9374 non-null   object 
 1   Application_ID              7598 non-null   object 
 2   Key                         9374 non-null   int64  
 3   Loan_Type                   9374 non-null   object 
 4   Contract_Start_Date         9374 non-null   object 
 5   Expected_Contract_End_Date  9374 non-null   object 
 6   Actual_Contract_End_Date    2593 non-null   object 
 7   Loan_Term                   9374 non-null   object 
 8   Total_Loan_Amount           9374 non-null   float64
dtypes: float64(1), int64(1), object(7)
memory usage: 659.2+ KB


**Observations:**

1. There are 9374 records of the data.

2. There are 9 features: Contract_ID, Application_ID, Key, Loan_Type, Contract_Start_Date, Expected_Contract_End_Date, Actual_Contract_End_Date Loan_Term, Total_Loan_Amount.

3. The data types are: float64(1), int64(1), object(7)

## **Snapshot Data**

In [ ]:
snapshot_data = pd.read_csv(file_path + "/3. ContractsSnapshotData.csv")
snapshot_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 274371 entries, 0 to 274370
Data columns (total 6 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   Contract_ID              274371 non-null  object 
 1   Snapshot Date            274371 non-null  object 
 2   Outstanding Amount       274371 non-null  float64
 3   Due Amount               274371 non-null  float64
 4   Number of Days Past Due  79158 non-null   float64
 5   Contract Status          274371 non-null  object 
dtypes: float64(3), object(3)
memory usage: 12.6+ MB


**Observations:**

1. There are 274371 records of the data.

2. There are 6 features: Contract_ID, Snapshot Date, Outstanding Amount, Due Amount, Number of Days Past Due, Contract Status.

3. The data types are: float64(3), object(3)

## **CRB Data**

In [ ]:
CRB_Data = pd.read_csv(file_path + "/4. CRB Data.csv")
CRB_Data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6750 entries, 0 to 6749
Data columns (total 4 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   Application_ID    6750 non-null   object
 1   Application_Date  6750 non-null   object
 2   CRB Score         6750 non-null   int64 
 3   CRB Grade         6750 non-null   object
dtypes: int64(1), object(3)
memory usage: 211.1+ KB


**Observations:**

1. There are 6750 records of the data.

2. There are 4 features:
       Application_ID, Application_Date, CRB Score, CRB Grade

3. The data types are: int64(1), object(3)

## **Current and Savings Account Data**

In [ ]:
account_data = pd.read_csv(file_path + "/5. Current and Savings Account Data.csv")
account_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 857677 entries, 0 to 857676
Data columns (total 31 columns):
 #   Column                                  Non-Null Count   Dtype  
---  ------                                  --------------   -----  
 0   ID                                      857677 non-null  int64  
 1   Account_Number                          857677 non-null  object 
 2   Application ID                          857672 non-null  object 
 3   Month (YYYY-MM)                         857647 non-null  object 
 4   Open Date                               857647 non-null  object 
 5   Closed Date                             161748 non-null  object 
 6   Average Balance                         857647 non-null  float64
 7   Overdraft Limit                         753484 non-null  float64
 8   Number of Days in Debit during Month    857647 non-null  float64
 9   Number of Days in Credit during Month   857647 non-null  float64
 10  Maximum End of Day Balance              8576

**Observations:**

1. There are 857677 records of the data.

2. There are 31 features: 'ID', 'Account_Number', 'Application ID', 'Month (YYYY-MM)',
       'Open Date', 'Closed Date', 'Average Balance', 'Overdraft Limit',
       'Number of Days in Debit during Month',
       'Number of Days in Credit during Month', 'Maximum End of Day Balance',
       'Minimum End of Day Balance', 'Sum of Monthly Debit Transactions',
       'Sum of Monthly Credit Transactions', 'Last Transaction Date',
       'Sum of ATM Deposits', 'Number of ATM Deposits',
       'Sum of ATM Withdrawals', 'Number of ATM Withdrawals',
       'Sum of Cheque Debit Transactions',
       'Number of Cheque Debit Transactions',
       'Sum of Bounced Cheque transactions', 'Number of Bounced Cheques',
       'Sum of Debit Card POS Transactions',
       'Number of Debit Card POS Transactions',
       'Sum of Over the Counter Withdrawals',
       'Number of Over the Counter withdrawals', 'Account Type',
       'Account Status', 'Number of Joint Holders', 'SpshotMonth'

3. The data types are:  float64(21), int64(1), object(9)

## **Important Features**

1. ApplicationData.csv: Loan_Approval_Status, Monthly_Income, Number_of_Dependents, Time_at_Current_Employment_(Months), Employment_Status
2. ContractsData.csv: Loan_Type, Loan_Term, Total_Loan_Amount
3. ContractsSnapshotData.csv: Outstanding Amount, Number of Days Past Due, Contract Status
4. CRB Data.csv: CRB Score, CRB Grade
5. Current and Savings Account Data.csv: Average Balance, Overdraft Limit, Sum of Monthly Credit Transactions, Number of Bounced Cheques

## **Data Cleaning / Preparation**

Understand the data

Pick the relevant columns

Handle missing data

Handle duplicates

Feature engineering

Merge datasets (if applicable)


### **Application Data**

In [ ]:
#The relevant columns
application_df = application_data[['Application_ID','Loan_Approval_Status','Age_at_Application','Marital_Status','Gender',
                                   'Own_Rent_Home','Occupation_Professional_Category','Type_of_Business_Industry_of_Employment','Employment_Status','Time_at_Current_Employment_(Months)',
                                   'Number_of_Dependents','Loan_Purpose','Monthly_Income','Government_Employee']]
application_df.head()

Application_ID Loan_Approval_Status  Age_at_Application Marital_Status  \
0     C302100001             Approved                39.0      Unmarried   
1     C302100002             Approved                29.0      Unmarried   
2     C302100003             Approved                24.0      Unmarried   
3     C302100004             Approved                54.0      Unmarried   
4     C302100005             Approved                30.0      Unmarried   

   Gender Own_Rent_Home Occupation_Professional_Category  \
0  Female           Own                              NaN   
1  Female        Family                              NaN   
2    Male        Family                              NaN   
3    Male          Rent                              NaN   
4  Female        Family                              NaN   

  Type_of_Business_Industry_of_Employment Employment_Status  \
0                                   Other         Full-Time   
1                                     NaN         Full-Time   
2                                     NaN         Full-Time   
3                                     NaN         Full-Time   
4                                     NaN         Full-Time   

   Time_at_Current_Employment_(Months)  Number_of_Dependents  \
0                                  6.0                   2.0   
1                                 25.0                   0.0   
2                                  7.0                   0.0   
3                                105.0                   4.0   
4                                 60.0                   1.0   

         Loan_Purpose  Monthly_Income Government_Employee  
0    Home Improvement         1500.59                 NaN  
1  Debt Consolidation         3769.36                 NaN  
2    Home Improvement         2197.10                 NaN  
3  Debt Consolidation         1094.45               False  
4            Vacation         3798.40                True

In [ ]:
# Finding Missing Values
missing_percentage = (application_df.isnull().sum() / len(application_df)) * 100
print("Missing Values Percentage:\n", missing_percentage)

Missing Values Percentage:
 Application_ID                              0.000000
Loan_Approval_Status                        1.453856
Age_at_Application                          0.221239
Marital_Status                             18.752634
Gender                                      0.042141
Own_Rent_Home                              13.737885
Occupation_Professional_Category            1.380110
Type_of_Business_Industry_of_Employment    24.041298
Employment_Status                           1.422250
Time_at_Current_Employment_(Months)         2.444164
Number_of_Dependents                       19.110830
Loan_Purpose                                0.442478
Monthly_Income                              0.000000
Government_Employee                         1.906869
dtype: float64


In [ ]:
# Drop rows with any missing values from application_df
clean_application_data = application_df.dropna()

# Display the number of remaining records
print("Number of records after dropping missing values:", clean_application_data.shape[0])

# Display cleaned dataset
print("The data without any missing values:\n", clean_application_data.isnull().sum())


Number of records after dropping missing values: 6353
The data without any missing values:
 Application_ID                             0
Loan_Approval_Status                       0
Age_at_Application                         0
Marital_Status                             0
Gender                                     0
Own_Rent_Home                              0
Occupation_Professional_Category           0
Type_of_Business_Industry_of_Employment    0
Employment_Status                          0
Time_at_Current_Employment_(Months)        0
Number_of_Dependents                       0
Loan_Purpose                               0
Monthly_Income                             0
Government_Employee                        0
dtype: int64


In [ ]:
# Checking for duplicates
print("No. of duplicates records in the dataset : ",clean_application_data.duplicated().sum())

No. of duplicates records in the dataset :  0


In [ ]:
clean_application_data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 6353 entries, 9 to 9490
Data columns (total 14 columns):
 #   Column                                   Non-Null Count  Dtype  
---  ------                                   --------------  -----  
 0   Application_ID                           6353 non-null   object 
 1   Loan_Approval_Status                     6353 non-null   object 
 2   Age_at_Application                       6353 non-null   float64
 3   Marital_Status                           6353 non-null   object 
 4   Gender                                   6353 non-null   object 
 5   Own_Rent_Home                            6353 non-null   object 
 6   Occupation_Professional_Category         6353 non-null   object 
 7   Type_of_Business_Industry_of_Employment  6353 non-null   object 
 8   Employment_Status                        6353 non-null   object 
 9   Time_at_Current_Employment_(Months)      6353 non-null   float64
 10  Number_of_Dependents                     6353 non-nul

**Observations:**

1. There are 7411 records of the data.

2. There are 4 features:
       'Application_ID', 'Loan_Approval_Status', 'Employment_Status',
       'Time_at_Current_Employment_(Months)', 'Number_of_Dependents',
       'Monthly_Income'

3. The data types are: float64(3), object(3)

### **ContractsData**

In [ ]:
# Selecting only the specified columns from the ContractsData since the application_data and Contracts id is different from the rest of the data
contracts_df = contracts_data[['Application_ID','Contract_ID','Key','Contract_Start_Date','Expected_Contract_End_Date',
                               'Actual_Contract_End_Date','Loan_Type','Loan_Term', 'Total_Loan_Amount']]
#'Contract_Start_Date','Expected_Contract_End_Date','Actual_Contract_End_Date','Loan_Type',
# Displaying the first few rows
contracts_df.head()

Application_ID Contract_ID  Key Contract_Start_Date  \
0     C302103426  CONT100001    1           4/12/2018   
1     C302103426  CONT100002    2           4/12/2018   
2     C302107460  CONT100003    1         202-1-08-31   
3     C302107453  CONT100004    1         202-1-08-28   
4     C302107447  CONT100005    1         202-1-08-27   

  Expected_Contract_End_Date Actual_Contract_End_Date  \
0                  4/12/2023                      NaN   
1                  4/12/2023                      NaN   
2                   9/1/2024                      NaN   
3                  8/28/2023                      NaN   
4                  8/27/2023                      NaN   

                 Loan_Type      Loan_Term  Total_Loan_Amount  
0  Unsecured Personal Loan  60 - Month(s)           24529.89  
1  Unsecured Personal Loan  60 - Month(s)           35085.79  
2  Unsecured Personal Loan  60 - Month(s)           53564.88  
3  Unsecured Personal Loan  48 - Month(s)           29073.86  
4  Unsecured Personal Loan  48 - Month(s)           11723.33

In [ ]:
contracts_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9374 entries, 0 to 9373
Data columns (total 9 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   Application_ID              7598 non-null   object 
 1   Contract_ID                 9374 non-null   object 
 2   Key                         9374 non-null   int64  
 3   Contract_Start_Date         9374 non-null   object 
 4   Expected_Contract_End_Date  9374 non-null   object 
 5   Actual_Contract_End_Date    2593 non-null   object 
 6   Loan_Type                   9374 non-null   object 
 7   Loan_Term                   9374 non-null   object 
 8   Total_Loan_Amount           9374 non-null   float64
dtypes: float64(1), int64(1), object(7)
memory usage: 659.2+ KB


In [ ]:
# Finding Missing Values
missing_percentage = (contracts_df.isnull().sum() / len(contracts_df)) * 100
print("Missing Values Percentage:\n", missing_percentage)

Missing Values Percentage:
 Application_ID                18.946021
Contract_ID                    0.000000
Key                            0.000000
Contract_Start_Date            0.000000
Expected_Contract_End_Date     0.000000
Actual_Contract_End_Date      72.338383
Loan_Type                      0.000000
Loan_Term                      0.000000
Total_Loan_Amount              0.000000
dtype: float64


In [ ]:
contracts_df['Actual_Contract_End_Date'].fillna(0, inplace = True)

In [ ]:
# Drop rows with any missing values from application_df
clean_contracts_data = contracts_df.dropna(subset=['Application_ID'])

clean_contracts_data.isnull().sum()

Application_ID                0
Contract_ID                   0
Key                           0
Contract_Start_Date           0
Expected_Contract_End_Date    0
Actual_Contract_End_Date      0
Loan_Type                     0
Loan_Term                     0
Total_Loan_Amount             0
dtype: int64

In [ ]:
# Checking for duplicates
print("No. of duplicates records in the dataset : ",clean_contracts_data.duplicated().sum())

No. of duplicates records in the dataset :  0


In [ ]:
clean_contracts_data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 7598 entries, 0 to 7597
Data columns (total 9 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   Application_ID              7598 non-null   object 
 1   Contract_ID                 7598 non-null   object 
 2   Key                         7598 non-null   int64  
 3   Contract_Start_Date         7598 non-null   object 
 4   Expected_Contract_End_Date  7598 non-null   object 
 5   Actual_Contract_End_Date    7598 non-null   object 
 6   Loan_Type                   7598 non-null   object 
 7   Loan_Term                   7598 non-null   object 
 8   Total_Loan_Amount           7598 non-null   float64
dtypes: float64(1), int64(1), object(7)
memory usage: 593.6+ KB


**Observations:**

1. There are 7598 records of the data.
2. There are 5 features:
       'Application_ID', 'Contract_ID', 'Loan_Type', 'Loan_Term','Total_Loan_Amount'
3. The data types are: float64(1), object(4)

### **SnapshotData**

In [ ]:
snapshot_df = snapshot_data[['Contract_ID','Outstanding Amount','Due Amount','Number of Days Past Due']]
snapshot_df.head()

Contract_ID  Outstanding Amount  Due Amount  Number of Days Past Due
0  CONT109375                0.00         0.0                      NaN
1  CONT109376              215.08         0.0                      NaN
2  CONT109376              215.08         0.0                      NaN
3  CONT109376              215.08         0.0                      NaN
4  CONT109376              215.08         0.0                      NaN

In [ ]:
snapshot_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 274371 entries, 0 to 274370
Data columns (total 4 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   Contract_ID              274371 non-null  object 
 1   Outstanding Amount       274371 non-null  float64
 2   Due Amount               274371 non-null  float64
 3   Number of Days Past Due  79158 non-null   float64
dtypes: float64(3), object(1)
memory usage: 8.4+ MB


In [ ]:
# Finding Missnapshot_dfsing Values
missing_percentage = (snapshot_df.isnull().sum() / len(snapshot_df)) * 100
print("Missing Values Percentage:\n", missing_percentage)

Missing Values Percentage:
 Contract_ID                 0.000000
Outstanding Amount          0.000000
Due Amount                  0.000000
Number of Days Past Due    71.149283
dtype: float64


In [ ]:
# Fill numerical columns with median (better for skewed data)
numerical_cols = snapshot_df.select_dtypes(include=['number']).columns
for col in numerical_cols:
    snapshot_df[col].fillna(snapshot_df[col].median(), inplace=True)

# Check if there are any remaining missing values
snapshot_df.isnull().sum()

Contract_ID                0
Outstanding Amount         0
Due Amount                 0
Number of Days Past Due    0
dtype: int64

In [ ]:
# Fill categorical columns with mode (most frequent value)
#categorical_cols = snapshot_df.select_dtypes(include=['object']).columns
#for col in categorical_cols:
    #snapshot_df[col].fillna(snapshot_df[col].mode()[0], inplace=True)

# Fill numerical columns with median (better for skewed data)
#numerical_cols = snapshot_df.select_dtypes(include=['number']).columns
#for col in numerical_cols:
    #snapshot_df[col].fillna(snapshot_df[col].median(), inplace=True)

# Check if there are any remaining missing values
#snapshot_df.isnull().sum()

In [ ]:
# Checking for duplicates
print("No. of duplicates records in the dataset : ",clean_snapshot_data.duplicated().sum())

No. of duplicates records in the dataset :  0


In [ ]:
print("Duplicates records in the dataset:", clean_snapshot_data.drop_duplicates(inplace=True))

Duplicates records in the dataset: None


In [ ]:
clean_snapshot_data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 73185 entries, 5 to 274347
Data columns (total 4 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Contract_ID              73185 non-null  object 
 1   Outstanding Amount       73185 non-null  float64
 2   Due Amount               73185 non-null  float64
 3   Number of Days Past Due  73185 non-null  float64
dtypes: float64(3), object(1)
memory usage: 2.8+ MB


**Observations:**
1. There are 257842 records of the data.
2. There are 4 features:
       'Contract_ID', 'Outstanding Amount', 'Number of Days Past Due',
       'Contract Status'
3. The data types are: float64(2), object(2)

### **CRB Data**

In [ ]:
CRB_df = CRB_Data[['Application_ID','CRB Score','CRB Grade']]
CRB_df.head()

Application_ID  CRB Score CRB Grade
0     C302102137        593         D
1     C302106175        625         C
2     C302106175        616         C
3     C302103572        629         C
4     C302100598        685         A

In [ ]:
# Finding Missing Values
missing_percentage = (CRB_df.isnull().sum() / len(CRB_df)) * 100
print("Missing Values Percentage:\n", missing_percentage)

Missing Values Percentage:
 Application_ID    0.0
CRB Score         0.0
CRB Grade         0.0
dtype: float64


In [ ]:
# Checking for duplicates
print("No. of duplicates records in the dataset : ",CRB_df.duplicated().sum())

No. of duplicates records in the dataset :  75


In [ ]:
print("Duplicates records in the dataset:", CRB_df.drop_duplicates(inplace=True))

Duplicates records in the dataset: None


In [ ]:
CRB_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 6675 entries, 0 to 6749
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   Application_ID  6675 non-null   object
 1   CRB Score       6675 non-null   int64 
 2   CRB Grade       6675 non-null   object
dtypes: int64(1), object(2)
memory usage: 208.6+ KB


**Observations:**
1. There are 6675 records of the data.
2. There are 2 features:
       'Application_ID', 'CRB Score', 'CRB Grade'
3. The data types are: int64(1), object(2)

### **Current and Savings Account Data**

In [ ]:
account_df = account_data[['Application ID','Average Balance', 'Overdraft Limit','Number of Days in Debit during Month',
                           'Sum of Monthly Credit Transactions','Sum of Monthly Debit Transactions', 'Sum of ATM Withdrawals',
                           'Number of ATM Withdrawals','Sum of Cheque Debit Transactions','Sum of Bounced Cheque transactions','Number of Bounced Cheques',
                           'Sum of Debit Card POS Transactions','Number of Debit Card POS Transactions','Sum of Over the Counter Withdrawals','Number of Over the Counter withdrawals',
                           'Account Type','Account Status','Number of Joint Holders','SpshotMonth']]

#Number of Days in Debit during Month
#Number of Days in Credit during Month

#Number of ATM Deposits
#Number of Cheque Debit Transactions
#Account Status



#Sum of ATM Deposits
#Number of ATM Deposits
#Sum of ATM Withdrawals
#Number of ATM Withdrawals
#Sum of Cheque Debit Transactions

#Sum of Monthly Debit Transactions
#Sum of Bounced Cheque transactions
#Account Type
#Number of Joint Holders
#SnapshotMonth

account_df.head()

Application ID  Average Balance  Overdraft Limit  \
0     C302100001          1986.09              0.0   
1     C302100001            23.45              NaN   
2     C302100001          8120.71              0.0   
3     C302100001            25.26              NaN   
4     C302100001          6624.10              0.0   

   Number of Days in Debit during Month  Sum of Monthly Credit Transactions  \
0                                   0.0                             9685.72   
1                                   0.0                               23.45   
2                                   0.0                                0.00   
3                                   0.0                               56.12   
4                                   0.0                                0.00   

   Sum of Monthly Debit Transactions  Sum of ATM Withdrawals  \
0                            1473.04                 1336.46   
1                                NaN                     NaN   
2                             708.09                     NaN   
3                               0.00                     NaN   
4                            1580.64                     NaN   

   Number of ATM Withdrawals  Sum of Cheque Debit Transactions  \
0                        3.0                               NaN   
1                        0.0                               NaN   
2                        0.0                               NaN   
3                        0.0                               NaN   
4                        0.0                               NaN   

   Sum of Bounced Cheque transactions  Number of Bounced Cheques  \
0                                 NaN                        0.0   
1                                 NaN                        0.0   
2                                 NaN                        0.0   
3                                 NaN                        0.0   
4                                 NaN                        0.0   

   Sum of Debit Card POS Transactions  Number of Debit Card POS Transactions  \
0                                 NaN                                    0.0   
1                                 NaN                                    0.0   
2                               58.62                                    1.0   
3                                 NaN                                    0.0   
4                              989.57                                    3.0   

   Sum of Over the Counter Withdrawals  \
0                                  NaN   
1                                  NaN   
2                                  NaN   
3                                  NaN   
4                                  NaN   

   Number of Over the Counter withdrawals Account Type Account Status  \
0                                     0.0      Savings         Active   
1                                     0.0      Savings         Closed   
2                                     0.0      Savings         Active   
3                                     0.0      Savings         Closed   
4                                     0.0      Savings         Active   

   Number of Joint Holders SpshotMonth  
0                      0.0  2015-09-01  
1                      0.0  2015-09-01  
2                      0.0  2015-10-01  
3                      0.0  2015-10-01  
4                      0.0  2015-11-01

In [ ]:
account_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 857677 entries, 0 to 857676
Data columns (total 19 columns):
 #   Column                                  Non-Null Count   Dtype  
---  ------                                  --------------   -----  
 0   Application ID                          857672 non-null  object 
 1   Average Balance                         857647 non-null  float64
 2   Overdraft Limit                         753484 non-null  float64
 3   Number of Days in Debit during Month    857647 non-null  float64
 4   Sum of Monthly Credit Transactions      857529 non-null  float64
 5   Sum of Monthly Debit Transactions       832991 non-null  float64
 6   Sum of ATM Withdrawals                  188709 non-null  float64
 7   Number of ATM Withdrawals               857647 non-null  float64
 8   Sum of Cheque Debit Transactions        634 non-null     float64
 9   Sum of Bounced Cheque transactions      2 non-null       float64
 10  Number of Bounced Cheques               8576

## **Feature Engineering**
Renamed the Application ID to Application_ID to be able to merge the dataframe(merged_df3) using a common key with the rest of the datafames(merged_df1 & merged_df3)

In [ ]:
# Rename column using rename()
current_savings_account = account_df.rename(columns={'Application ID': 'Application_ID'})
current_savings_account.head()

Application_ID  Average Balance  Overdraft Limit  \
0     C302100001          1986.09              0.0   
1     C302100001            23.45              NaN   
2     C302100001          8120.71              0.0   
3     C302100001            25.26              NaN   
4     C302100001          6624.10              0.0   

   Number of Days in Debit during Month  Sum of Monthly Credit Transactions  \
0                                   0.0                             9685.72   
1                                   0.0                               23.45   
2                                   0.0                                0.00   
3                                   0.0                               56.12   
4                                   0.0                                0.00   

   Sum of Monthly Debit Transactions  Sum of ATM Withdrawals  \
0                            1473.04                 1336.46   
1                                NaN                     NaN   
2                             708.09                     NaN   
3                               0.00                     NaN   
4                            1580.64                     NaN   

   Number of ATM Withdrawals  Sum of Cheque Debit Transactions  \
0                        3.0                               NaN   
1                        0.0                               NaN   
2                        0.0                               NaN   
3                        0.0                               NaN   
4                        0.0                               NaN   

   Sum of Bounced Cheque transactions  Number of Bounced Cheques  \
0                                 NaN                        0.0   
1                                 NaN                        0.0   
2                                 NaN                        0.0   
3                                 NaN                        0.0   
4                                 NaN                        0.0   

   Sum of Debit Card POS Transactions  Number of Debit Card POS Transactions  \
0                                 NaN                                    0.0   
1                                 NaN                                    0.0   
2                               58.62                                    1.0   
3                                 NaN                                    0.0   
4                              989.57                                    3.0   

   Sum of Over the Counter Withdrawals  \
0                                  NaN   
1                                  NaN   
2                                  NaN   
3                                  NaN   
4                                  NaN   

   Number of Over the Counter withdrawals Account Type Account Status  \
0                                     0.0      Savings         Active   
1                                     0.0      Savings         Closed   
2                                     0.0      Savings         Active   
3                                     0.0      Savings         Closed   
4                                     0.0      Savings         Active   

   Number of Joint Holders SpshotMonth  
0                      0.0  2015-09-01  
1                      0.0  2015-09-01  
2                      0.0  2015-10-01  
3                      0.0  2015-10-01  
4                      0.0  2015-11-01

In [ ]:
# Finding Missing Values
missing_percentage = (current_savings_account.isnull().sum() / len(current_savings_account)) * 100
print("Missing Values Percentage:\n", missing_percentage)

Missing Values Percentage:
 Application_ID                             0.000583
Average Balance                            0.003498
Overdraft Limit                           12.148280
Number of Days in Debit during Month       0.003498
Sum of Monthly Credit Transactions         0.017256
Sum of Monthly Debit Transactions          2.878240
Sum of ATM Withdrawals                    77.997661
Number of ATM Withdrawals                  0.003498
Sum of Cheque Debit Transactions          99.926079
Sum of Bounced Cheque transactions        99.999767
Number of Bounced Cheques                  0.003498
Sum of Debit Card POS Transactions        86.628766
Number of Debit Card POS Transactions      0.003498
Sum of Over the Counter Withdrawals       92.530521
Number of Over the Counter withdrawals     0.003498
Account Type                               0.003498
Account Status                             0.003498
Number of Joint Holders                    0.003498
SpshotMonth                         

In [ ]:
import pandas as pd

# Replace missing values in categorical column ('Application_ID') with its mode
current_savings_account['Application_ID'] = current_savings_account['Application_ID'].fillna(current_savings_account['Application_ID'].mode()[0])

# Identify numerical columns
numeric_cols = current_savings_account.select_dtypes(include=['number']).columns

# Replace missing values in numerical columns with their median
current_savings_account[numeric_cols] = current_savings_account[numeric_cols].fillna(current_savings_account[numeric_cols].median())

#dropping missing values from specific columns
df_cleaned = current_savings_account.dropna(subset=['Account Type', 'Account Status','SpshotMonth'])

# Store the cleaned dataframe
clean_account_df = df_cleaned

# Check for remaining missing values
print(clean_account_df.isnull().sum())

Application_ID                            0
Average Balance                           0
Overdraft Limit                           0
Number of Days in Debit during Month      0
Sum of Monthly Credit Transactions        0
Sum of Monthly Debit Transactions         0
Sum of ATM Withdrawals                    0
Number of ATM Withdrawals                 0
Sum of Cheque Debit Transactions          0
Sum of Bounced Cheque transactions        0
Number of Bounced Cheques                 0
Sum of Debit Card POS Transactions        0
Number of Debit Card POS Transactions     0
Sum of Over the Counter Withdrawals       0
Number of Over the Counter withdrawals    0
Account Type                              0
Account Status                            0
Number of Joint Holders                   0
SpshotMonth                               0
dtype: int64


In [ ]:
# Checking for duplicates
print("No. of duplicates records in the dataset : ", clean_account_df.duplicated().sum())

No. of duplicates records in the dataset :  2615


In [ ]:
# Checking for dropping duplicates
print("No. of records after dropping duplicates in the dataset : ", clean_account_df.drop_duplicates(inplace = True))

No. of records after dropping duplicates in the dataset :  None


In [ ]:
clean_account_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 855032 entries, 0 to 857671
Data columns (total 19 columns):
 #   Column                                  Non-Null Count   Dtype  
---  ------                                  --------------   -----  
 0   Application_ID                          855032 non-null  object 
 1   Average Balance                         855032 non-null  float64
 2   Overdraft Limit                         855032 non-null  float64
 3   Number of Days in Debit during Month    855032 non-null  float64
 4   Sum of Monthly Credit Transactions      855032 non-null  float64
 5   Sum of Monthly Debit Transactions       855032 non-null  float64
 6   Sum of ATM Withdrawals                  855032 non-null  float64
 7   Number of ATM Withdrawals               855032 non-null  float64
 8   Sum of Cheque Debit Transactions        855032 non-null  float64
 9   Sum of Bounced Cheque transactions      855032 non-null  float64
 10  Number of Bounced Cheques               855032 no

**Observations:**
1. There are 587607 records of the data.
2. There are 5 features:
       'Application_ID', 'Average Balance', 'Overdraft Limit',
       'Sum of Monthly Credit Transactions', 'Number of Bounced Cheques'
3. The data types are: float64(4), object(1)

##  Merge1 Datasets: ApplicationData , CRB Data and ContractsData

#### The first merge being the ApplicationData, CRB Data and ContractsData , using:
1. Application_ID which is a common key
2. Inner joint used to combine rows from two or more tables based on a matching condition. It returns only the rows where there is a match in both tables.

In [ ]:
import pandas as pd


# First, merge application_data with contracts_data
merged_df = pd.merge(clean_application_data, clean_contracts_data, on='Application_ID', how='inner')

# Then, merge the result with CRB_df
merged_df = pd.merge(merged_df, CRB_df, on='Application_ID', how='inner')

# Display the first few rows of the merged dataset
merged_df.head()

Application_ID Loan_Approval_Status  Age_at_Application Marital_Status  \
0     C302100023             Approved                58.0        Widowed   
1     C302100023             Approved                58.0        Widowed   
2     C302100026             Approved                61.0        Married   
3     C302100027             Approved                62.0        Married   
4     C302100036             Approved                31.0      Unmarried   

   Gender Own_Rent_Home Occupation_Professional_Category  \
0  Female           Own                        Pensioner   
1  Female           Own                        Pensioner   
2    Male           Own                         Employee   
3    Male           Own                     Professional   
4    Male        Family         Service and Sales Worker   

  Type_of_Business_Industry_of_Employment Employment_Status  \
0                       Retiree/Pensioner         Full-Time   
1                       Retiree/Pensioner         Full-Time   
2                              Unemployed          Contract   
3                           Civil Service           Retired   
4                                 Unknown         Full-Time   

   Time_at_Current_Employment_(Months)  ...  Contract_ID Key  \
0                                120.0  ...   CONT106977   1   
1                                120.0  ...   CONT106977   1   
2                                 23.0  ...   CONT107322   1   
3                                505.0  ...   CONT102493   1   
4                                156.0  ...   CONT107263   1   

   Contract_Start_Date Expected_Contract_End_Date Actual_Contract_End_Date  \
0            5/17/2017                  5/17/2018                5/28/2018   
1            5/17/2017                  5/17/2018                5/28/2018   
2            7/25/2017                  7/31/2022                        0   
3            7/30/2017                  8/10/2021                        0   
4            9/21/2017                  9/30/2022              202-1-02-12   

                 Loan_Type      Loan_Term Total_Loan_Amount CRB Score  \
0  Unsecured Personal Loan  12 - Month(s)           2344.67       999   
1  Unsecured Personal Loan  12 - Month(s)           2344.67       686   
2  Unsecured Personal Loan  60 - Month(s)          21101.99       708   
3  Unsecured Personal Loan  48 - Month(s)          23446.66       626   
4  Unsecured Personal Loan  60 - Month(s)          17233.29       644   

  CRB Grade  
0         X  
1         A  
2         A  
3         C  
4         C  

[5 rows x 24 columns]

In [ ]:
merged_df.isnull().sum()

Application_ID                             0
Loan_Approval_Status                       0
Age_at_Application                         0
Marital_Status                             0
Gender                                     0
Own_Rent_Home                              0
Occupation_Professional_Category           0
Type_of_Business_Industry_of_Employment    0
Employment_Status                          0
Time_at_Current_Employment_(Months)        0
Number_of_Dependents                       0
Loan_Purpose                               0
Monthly_Income                             0
Government_Employee                        0
Contract_ID                                0
Key                                        0
Contract_Start_Date                        0
Expected_Contract_End_Date                 0
Actual_Contract_End_Date                   0
Loan_Type                                  0
Loan_Term                                  0
Total_Loan_Amount                          0
CRB Score                                  0
CRB Grade                                  0
dtype: int64

In [ ]:
merged_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4819 entries, 0 to 4818
Data columns (total 24 columns):
 #   Column                                   Non-Null Count  Dtype  
---  ------                                   --------------  -----  
 0   Application_ID                           4819 non-null   object 
 1   Loan_Approval_Status                     4819 non-null   object 
 2   Age_at_Application                       4819 non-null   float64
 3   Marital_Status                           4819 non-null   object 
 4   Gender                                   4819 non-null   object 
 5   Own_Rent_Home                            4819 non-null   object 
 6   Occupation_Professional_Category         4819 non-null   object 
 7   Type_of_Business_Industry_of_Employment  4819 non-null   object 
 8   Employment_Status                        4819 non-null   object 
 9   Time_at_Current_Employment_(Months)      4819 non-null   float64
 10  Number_of_Dependents                     4819 no

**Observations:**

1. There are 5576 records of the data.

2. There are 12 features:
       ''Application_ID', 'Loan_Approval_Status', 'Employment_Status',
       'Time_at_Current_Employment_(Months)', 'Number_of_Dependents',
       'Monthly_Income', 'Contract_ID', 'Loan_Type', 'Loan_Term',
       'Total_Loan_Amount', 'CRB Score', 'CRB Grade'
3. The data types are: float64(4), int64(1), object(7)

##  Merge2 Datasets: Merged_df and ContractsSnapshotData

#### The second merge being the merged_df that comprises of the ApplicationData, CRB Data and ContractsData , with ContractsSnapshotData using:
1. Contract_ID which is a common key
2. Inner joint used to combine rows from two or more tables based on a matching condition. It returns only the rows where there is a match in both tables.

In [ ]:
# Then, merge the result with CRB_df
merged_df2 = pd.merge(merged_df, clean_snapshot_data , on='Contract_ID', how='inner')

# Display the first few rows of the merged dataset
merged_df2.head()

Application_ID Loan_Approval_Status  Age_at_Application Marital_Status  \
0     C302100023             Approved                58.0        Widowed   
1     C302100023             Approved                58.0        Widowed   
2     C302100036             Approved                31.0      Unmarried   
3     C302100036             Approved                31.0      Unmarried   
4     C302100036             Approved                31.0      Unmarried   

   Gender Own_Rent_Home Occupation_Professional_Category  \
0  Female           Own                        Pensioner   
1  Female           Own                        Pensioner   
2    Male        Family         Service and Sales Worker   
3    Male        Family         Service and Sales Worker   
4    Male        Family         Service and Sales Worker   

  Type_of_Business_Industry_of_Employment Employment_Status  \
0                       Retiree/Pensioner         Full-Time   
1                       Retiree/Pensioner         Full-Time   
2                                 Unknown         Full-Time   
3                                 Unknown         Full-Time   
4                                 Unknown         Full-Time   

   Time_at_Current_Employment_(Months)  ...  Expected_Contract_End_Date  \
0                                120.0  ...                   5/17/2018   
1                                120.0  ...                   5/17/2018   
2                                156.0  ...                   9/30/2022   
3                                156.0  ...                   9/30/2022   
4                                156.0  ...                   9/30/2022   

  Actual_Contract_End_Date                Loan_Type      Loan_Term  \
0                5/28/2018  Unsecured Personal Loan  12 - Month(s)   
1                5/28/2018  Unsecured Personal Loan  12 - Month(s)   
2              202-1-02-12  Unsecured Personal Loan  60 - Month(s)   
3              202-1-02-12  Unsecured Personal Loan  60 - Month(s)   
4              202-1-02-12  Unsecured Personal Loan  60 - Month(s)   

  Total_Loan_Amount  CRB Score CRB Grade Outstanding Amount Due Amount  \
0           2344.67        999         X               0.00     235.43   
1           2344.67        686         A               0.00     235.43   
2          17233.29        644         C           14283.42     427.60   
3          17233.29        644         C               0.00     427.60   
4          17233.29        642         C           14283.42     427.60   

  Number of Days Past Due  
0                    15.0  
1                    15.0  
2                     1.0  
3                    30.0  
4                     1.0  

[5 rows x 27 columns]

In [ ]:
merged_df2.isnull().sum()

Application_ID                             0
Loan_Approval_Status                       0
Age_at_Application                         0
Marital_Status                             0
Gender                                     0
Own_Rent_Home                              0
Occupation_Professional_Category           0
Type_of_Business_Industry_of_Employment    0
Employment_Status                          0
Time_at_Current_Employment_(Months)        0
Number_of_Dependents                       0
Loan_Purpose                               0
Monthly_Income                             0
Government_Employee                        0
Contract_ID                                0
Key                                        0
Contract_Start_Date                        0
Expected_Contract_End_Date                 0
Actual_Contract_End_Date                   0
Loan_Type                                  0
Loan_Term                                  0
Total_Loan_Amount                          0
CRB Score                                  0
CRB Grade                                  0
Outstanding Amount                         0
Due Amount                                 0
Number of Days Past Due                    0
dtype: int64

In [ ]:
merged_df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11273 entries, 0 to 11272
Data columns (total 27 columns):
 #   Column                                   Non-Null Count  Dtype  
---  ------                                   --------------  -----  
 0   Application_ID                           11273 non-null  object 
 1   Loan_Approval_Status                     11273 non-null  object 
 2   Age_at_Application                       11273 non-null  float64
 3   Marital_Status                           11273 non-null  object 
 4   Gender                                   11273 non-null  object 
 5   Own_Rent_Home                            11273 non-null  object 
 6   Occupation_Professional_Category         11273 non-null  object 
 7   Type_of_Business_Industry_of_Employment  11273 non-null  object 
 8   Employment_Status                        11273 non-null  object 
 9   Time_at_Current_Employment_(Months)      11273 non-null  float64
 10  Number_of_Dependents                     11273

**Observations:**

1. There are 83127 records of the data.

2. There are 15 features:
       'Application_ID', 'Loan_Approval_Status', 'Employment_Status',
       'Time_at_Current_Employment_(Months)', 'Number_of_Dependents',
       'Monthly_Income', 'Contract_ID', 'Loan_Type', 'Loan_Term',
       'Total_Loan_Amount', 'CRB Score', 'CRB Grade', 'Outstanding Amount',
       'Number of Days Past Due', 'Contract Status'
3. The data types are: float64(6), int64(1), object(8)

##  Merge3 Datasets: Merged_df2 and Current and Savings Account Data

#### The third merge being the merged_df2 that comprises of the ApplicationData, CRB Data, ContractsData and ContractsSnapshotData, with Current and Savings Account Data  using:
1. Application_ID which is a common key
2. Inner joint used to combine rows from two or more tables based on a matching condition. It returns only the rows where there is a match in both tables.

In [ ]:
# Then, merge the result with CRB_df
merged_df3 = pd.merge(merged_df2, clean_account_df , on='Application_ID', how='inner')

# Display the first few rows of the merged dataset
merged_df3.head()

Application_ID Loan_Approval_Status  Age_at_Application Marital_Status  \
0     C302100023             Approved                58.0        Widowed   
1     C302100023             Approved                58.0        Widowed   
2     C302100023             Approved                58.0        Widowed   
3     C302100023             Approved                58.0        Widowed   
4     C302100023             Approved                58.0        Widowed   

   Gender Own_Rent_Home Occupation_Professional_Category  \
0  Female           Own                        Pensioner   
1  Female           Own                        Pensioner   
2  Female           Own                        Pensioner   
3  Female           Own                        Pensioner   
4  Female           Own                        Pensioner   

  Type_of_Business_Industry_of_Employment Employment_Status  \
0                       Retiree/Pensioner         Full-Time   
1                       Retiree/Pensioner         Full-Time   
2                       Retiree/Pensioner         Full-Time   
3                       Retiree/Pensioner         Full-Time   
4                       Retiree/Pensioner         Full-Time   

   Time_at_Current_Employment_(Months)  ...  \
0                                120.0  ...   
1                                120.0  ...   
2                                120.0  ...   
3                                120.0  ...   
4                                120.0  ...   

   Sum of Bounced Cheque transactions Number of Bounced Cheques  \
0                            1232.205                       0.0   
1                            1232.205                       0.0   
2                            1232.205                       0.0   
3                            1232.205                       0.0   
4                            1232.205                       0.0   

   Sum of Debit Card POS Transactions Number of Debit Card POS Transactions  \
0                             260.695                                   0.0   
1                             260.695                                   0.0   
2                             260.695                                   0.0   
3                             260.695                                   0.0   
4                             260.695                                   0.0   

  Sum of Over the Counter Withdrawals  Number of Over the Counter withdrawals  \
0                              474.44                                     0.0   
1                              474.44                                     0.0   
2                              474.44                                     0.0   
3                              474.44                                     0.0   
4                              474.44                                     0.0   

  Account Type Account Status Number of Joint Holders SpshotMonth  
0      Savings         Closed                     1.0  2012-09-01  
1      Savings         Closed                     1.0  2012-10-01  
2      Savings         Closed                     1.0  2012-11-01  
3      Savings         Closed                     1.0  2012-12-01  
4      Savings         Closed                     1.0  2013-01-01  

[5 rows x 45 columns]

In [ ]:
merged_df3.isnull().sum()

Application_ID                             0
Loan_Approval_Status                       0
Age_at_Application                         0
Marital_Status                             0
Gender                                     0
Own_Rent_Home                              0
Occupation_Professional_Category           0
Type_of_Business_Industry_of_Employment    0
Employment_Status                          0
Time_at_Current_Employment_(Months)        0
Number_of_Dependents                       0
Loan_Purpose                               0
Monthly_Income                             0
Government_Employee                        0
Contract_ID                                0
Key                                        0
Contract_Start_Date                        0
Expected_Contract_End_Date                 0
Actual_Contract_End_Date                   0
Loan_Type                                  0
Loan_Term                                  0
Total_Loan_Amount                          0
CRB Score                                  0
CRB Grade                                  0
Outstanding Amount                         0
Due Amount                                 0
Number of Days Past Due                    0
Average Balance                            0
Overdraft Limit                            0
Number of Days in Debit during Month       0
Sum of Monthly Credit Transactions         0
Sum of Monthly Debit Transactions          0
Sum of ATM Withdrawals                     0
Number of ATM Withdrawals                  0
Sum of Cheque Debit Transactions           0
Sum of Bounced Cheque transactions         0
Number of Bounced Cheques                  0
Sum of Debit Card POS Transactions         0
Number of Debit Card POS Transactions      0
Sum of Over the Counter Withdrawals        0
Number of Over the Counter withdrawals     0
Account Type                               0
Account Status                             0
Number of Joint Holders                    0
SpshotMonth                                0
dtype: int64

In [ ]:
merged_df3.describe()

Age_at_Application  Time_at_Current_Employment_(Months)  \
count        1.095599e+06                         1.095599e+06   
mean         3.581358e+01                         8.583390e+01   
std          9.664335e+00                         7.821202e+01   
min          1.900000e+01                         1.000000e+00   
25%          2.800000e+01                         2.800000e+01   
50%          3.400000e+01                         5.800000e+01   
75%          4.300000e+01                         1.200000e+02   
max          6.700000e+01                         4.740000e+02   

       Number_of_Dependents  Monthly_Income           Key  Total_Loan_Amount  \
count          1.095599e+06    1.095599e+06  1.095599e+06       1.095599e+06   
mean           9.523375e-01    1.442349e+05  1.026113e+00       2.829901e+04   
std            1.118622e+00    3.793775e+06  1.646076e-01       2.459320e+04   
min            0.000000e+00    1.935800e+02  1.000000e+00       2.344670e+03   
25%            0.000000e+00    1.636380e+03  1.000000e+00       1.055100e+04   
50%            1.000000e+00    2.734840e+03  1.000000e+00       2.095306e+04   
75%            2.000000e+00    4.399960e+03  1.000000e+00       3.717006e+04   
max            9.000000e+00    2.056265e+08  3.000000e+00       1.906946e+05   

          CRB Score  Outstanding Amount    Due Amount  \
count  1.095599e+06        1.095599e+06  1.095599e+06   
mean   6.055477e+02        2.435275e+04  2.446376e+03   
std    1.364384e+02        2.350362e+04  4.700436e+03   
min    2.500000e+02        0.000000e+00  0.000000e+00   
25%    5.890000e+02        8.174630e+03  3.532000e+02   
50%    6.210000e+02        1.723700e+04  8.534100e+02   
75%    6.490000e+02        3.210415e+04  2.363140e+03   
max    9.990000e+02        1.806060e+05  1.038357e+05   

       Number of Days Past Due  ...  Sum of ATM Withdrawals  \
count             1.095599e+06  ...            1.095599e+06   
mean              9.732961e+01  ...            6.791401e+02   
std               1.712412e+02  ...            5.108248e+02   
min               1.000000e+00  ...            2.340000e+00   
25%               2.000000e+00  ...            6.260300e+02   
50%               2.500000e+01  ...            6.260300e+02   
75%               1.030000e+02  ...            6.260300e+02   
max               1.223000e+03  ...            2.684642e+04   

       Number of ATM Withdrawals  Sum of Cheque Debit Transactions  \
count               1.095599e+06                      1.095599e+06   
mean                8.968227e-01                      1.411778e+03   
std                 2.839823e+00                      5.653059e+02   
min                 0.000000e+00                      1.011900e+02   
25%                 0.000000e+00                      1.406800e+03   
50%                 0.000000e+00                      1.406800e+03   
75%                 0.000000e+00                      1.406800e+03   
max                 6.800000e+01                      1.340240e+05   

       Sum of Bounced Cheque transactions  Number of Bounced Cheques  \
count                        1.095599e+06               1.095599e+06   
mean                         1.232209e+03               3.650971e-06   
std                          2.125641e+00               1.910749e-03   
min                          1.232205e+03               0.000000e+00   
25%                          1.232205e+03               0.000000e+00   
50%                          1.232205e+03               0.000000e+00   
75%                          1.232205e+03               0.000000e+00   
max                          2.344670e+03               1.000000e+00   

       Sum of Debit Card POS Transactions  \
count                        1.095599e+06   
mean                         2.947998e+02   
std                          3.928006e+02   
min                          0.000000e+00   
25%                          2.606950e+02   
50%                          2.606950e+02   
75%           

In [ ]:
merged_df3.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1095599 entries, 0 to 1095598
Data columns (total 45 columns):
 #   Column                                   Non-Null Count    Dtype  
---  ------                                   --------------    -----  
 0   Application_ID                           1095599 non-null  object 
 1   Loan_Approval_Status                     1095599 non-null  object 
 2   Age_at_Application                       1095599 non-null  float64
 3   Marital_Status                           1095599 non-null  object 
 4   Gender                                   1095599 non-null  object 
 5   Own_Rent_Home                            1095599 non-null  object 
 6   Occupation_Professional_Category         1095599 non-null  object 
 7   Type_of_Business_Industry_of_Employment  1095599 non-null  object 
 8   Employment_Status                        1095599 non-null  object 
 9   Time_at_Current_Employment_(Months)      1095599 non-null  float64
 10  Number_of_Dependen

In [ ]:
# Finding Missing Values
missing_percentage = (merged_df3.isnull().sum() / len(merged_df3)) * 100
print("Missing Values Percentage:\n", missing_percentage)

Missing Values Percentage:
 Application_ID                             0.0
Loan_Approval_Status                       0.0
Age_at_Application                         0.0
Marital_Status                             0.0
Gender                                     0.0
Own_Rent_Home                              0.0
Occupation_Professional_Category           0.0
Type_of_Business_Industry_of_Employment    0.0
Employment_Status                          0.0
Time_at_Current_Employment_(Months)        0.0
Number_of_Dependents                       0.0
Loan_Purpose                               0.0
Monthly_Income                             0.0
Government_Employee                        0.0
Contract_ID                                0.0
Key                                        0.0
Contract_Start_Date                        0.0
Expected_Contract_End_Date                 0.0
Actual_Contract_End_Date                   0.0
Loan_Type                                  0.0
Loan_Term                       

In [ ]:
merged_df3.drop(['Application_ID', 'Contract_ID'], axis=1, inplace=True)
merged_df3.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1095599 entries, 0 to 1095598
Data columns (total 43 columns):
 #   Column                                   Non-Null Count    Dtype  
---  ------                                   --------------    -----  
 0   Loan_Approval_Status                     1095599 non-null  object 
 1   Age_at_Application                       1095599 non-null  float64
 2   Marital_Status                           1095599 non-null  object 
 3   Gender                                   1095599 non-null  object 
 4   Own_Rent_Home                            1095599 non-null  object 
 5   Occupation_Professional_Category         1095599 non-null  object 
 6   Type_of_Business_Industry_of_Employment  1095599 non-null  object 
 7   Employment_Status                        1095599 non-null  object 
 8   Time_at_Current_Employment_(Months)      1095599 non-null  float64
 9   Number_of_Dependents                     1095599 non-null  float64
 10  Loan_Purpose      

In [ ]:

print("The number of records are : ", merged_df3.shape[0])
print("The number of features are : ", merged_df3.shape[1])
print("The list of features is : ", merged_df3.columns)

The number of records are :  1095599
The number of features are :  43
The list of features is :  Index(['Loan_Approval_Status', 'Age_at_Application', 'Marital_Status',
       'Gender', 'Own_Rent_Home', 'Occupation_Professional_Category',
       'Type_of_Business_Industry_of_Employment', 'Employment_Status',
       'Time_at_Current_Employment_(Months)', 'Number_of_Dependents',
       'Loan_Purpose', 'Monthly_Income', 'Government_Employee', 'Key',
       'Contract_Start_Date', 'Expected_Contract_End_Date',
       'Actual_Contract_End_Date', 'Loan_Type', 'Loan_Term',
       'Total_Loan_Amount', 'CRB Score', 'CRB Grade', 'Outstanding Amount',
       'Due Amount', 'Number of Days Past Due', 'Average Balance',
       'Overdraft Limit', 'Number of Days in Debit during Month',
       'Sum of Monthly Credit Transactions',
       'Sum of Monthly Debit Transactions', 'Sum of ATM Withdrawals',
       'Number of ATM Withdrawals', 'Sum of Cheque Debit Transactions',
       'Sum of Bounced Cheque tra

**Observations:**

1. There are 6596471 records of the data.

2. There are 17 features:
       'Loan_Approval_Status', 'Employment_Status',
       'Time_at_Current_Employment_(Months)', 'Number_of_Dependents',
       'Monthly_Income', 'Loan_Type', 'Loan_Term', 'Total_Loan_Amount',
       'CRB Score', 'CRB Grade', 'Outstanding Amount',
       'Number of Days Past Due', 'Contract Status', 'Average Balance',
       'Overdraft Limit', 'Sum of Monthly Credit Transactions',
       'Number of Bounced Cheques'

3. The data types are: float64(10), int64(1), object(6)

In [ ]:
# Save DataFrame to a CSV file
#final_data = combined_df.to_csv("CRB_DATA.csv", index=False)
#print("DataFrame saved as 'CRB_DATA.csv'")

In [ ]:
#file_path = "/content/drive/MyDrive/CRB_Data"

#final_data = pd.read_csv(file_path + "/CRB_DATA.csv")
# Checking the features and no. of records in the dataset

#print("The number of records are : ", final_data.shape[0])
#print("The number of features are : ", final_data.shape[1])
#print("The list of features is : ", final_data.columns)

In [ ]:
#final_data.info()